File: server/service/src/sync/translations/stock_line.rs
use crate::sync::translations::{
    barcode::BarcodeTranslation, campaign::CampaignTranslation, item::ItemTranslation,
    item_variant::ItemVariantTranslation, location::LocationTranslation, name::NameTranslation,
    store::StoreTranslation, vvm_status::VVMStatusTranslation,
};

use chrono::NaiveDate;
use repository::{
    campaign_row::CampaignRowRepository, item_variant::item_variant_row::ItemVariantRowRepository,
    vvm_status::vvm_status_row::VVMStatusRowRepository, BarcodeRowRepository, ChangelogRow,
    ChangelogTableName, EqualFilter, LocationRowRepository, ProgramRowRepository, StockLine,
    StockLineFilter, StockLineRepository, StockLineRow, StorageConnection, SyncBufferRow,
};
use serde::{Deserialize, Serialize};
use util::sync_serde::{
    date_option_to_isostring, empty_str_as_option_string, object_fields_as_option,
    zero_date_as_option,
};

use super::{utils::clear_invalid_fk, PullTranslateResult, PushTranslateResult, SyncTranslation};

const RECORD_TABLE: &str = "stock_line";

#[allow(non_snake_case)]
#[derive(Deserialize, Serialize, Default)]
pub struct StockLineRowOmsFields {
    #[serde(default)]
    #[serde(deserialize_with = "empty_str_as_option_string")]
    pub campaign_id: Option<String>,
    #[serde(default)]
    #[serde(deserialize_with = "empty_str_as_option_string")]
    pub program_id: Option<String>,
    #[serde(default)]
    pub manufacture_date: Option<NaiveDate>,
}

#[allow(non_snake_case)]
#[derive(Deserialize, Serialize)]
pub struct LegacyStockLineRow {
    pub ID: String,
    pub store_ID: String,
    pub item_ID: String,
    #[serde(deserialize_with = "empty_str_as_option_string")]
    pub batch: Option<String>,
    #[serde(deserialize_with = "zero_date_as_option")]
    #[serde(serialize_with = "date_option_to_isostring")]
    pub expiry_date: Option<NaiveDate>,
    pub hold: bool,
    #[serde(deserialize_with = "empty_str_as_option_string")]
    pub location_ID: Option<String>,
    pub pack_size: f64,
    pub available: f64,
    pub quantity: f64,
    pub cost_price: f64,
    pub sell_price: f64,
    #[serde(deserialize_with = "empty_str_as_option_string")]
    pub note: Option<String>,
    #[serde(rename = "name_ID")]
    #[serde(deserialize_with = "empty_str_as_option_string")]
    pub supplier_id: Option<String>,
    #[serde(deserialize_with = "empty_str_as_option_string", rename = "barcodeID")]
    pub barcode_id: Option<String>,
    #[serde(rename = "om_item_variant_id")]
    #[serde(deserialize_with = "empty_str_as_option_string")]
    #[serde(default)]
    pub item_variant_id: Option<String>,
    #[serde(default)]
    #[serde(deserialize_with = "empty_str_as_option_string")]
    pub donor_id: Option<String>,
    #[serde(deserialize_with = "empty_str_as_option_string")]
    pub vvm_status_id: Option<String>,
    #[serde(rename = "manufacturer_ID")]
    #[serde(deserialize_with = "empty_str_as_option_string")]
    #[serde(default)]
    pub manufacturer_id: Option<String>,
    #[serde(default)]
    #[serde(deserialize_with = "object_fields_as_option")]
    pub oms_fields: Option<StockLineRowOmsFields>,
    pub total_volume: f64,
    pub volume_per_pack: f64,
}
// Needs to be added to all_translators()
#[deny(dead_code)]
pub(crate) fn boxed() -> Box<dyn SyncTranslation> {
    Box::new(StockLineTranslation)
}

pub(super) struct StockLineTranslation;
impl SyncTranslation for StockLineTranslation {
    fn table_name(&self) -> &str {
        "item_line"
    }

    fn pull_dependencies(&self) -> Vec<&str> {
        vec![
            ItemTranslation.table_name(),
            ItemVariantTranslation.table_name(),
            NameTranslation.table_name(),
            StoreTranslation.table_name(),
            LocationTranslation.table_name(),
            BarcodeTranslation.table_name(),
            VVMStatusTranslation.table_name(),
            CampaignTranslation.table_name(),
        ]
    }

    fn change_log_type(&self) -> Option<ChangelogTableName> {
        Some(ChangelogTableName::StockLine)
    }

    fn try_translate_from_upsert_sync_record(
        &self,
        connection: &StorageConnection,
        sync_record: &SyncBufferRow,
    ) -> Result<PullTranslateResult, anyhow::Error> {
        let LegacyStockLineRow {
            ID,
            store_ID,
            item_ID,
            batch,
            expiry_date,
            hold,
            location_ID,
            pack_size,
            available,
            quantity,
            cost_price,
            sell_price,
            note,
            supplier_id,
            barcode_id,
            item_variant_id,
            donor_id,
            vvm_status_id,
            manufacturer_id,
            oms_fields,
            total_volume,
            volume_per_pack,
        } = serde_json::from_str::<LegacyStockLineRow>(&sync_record.data)?;

        let barcode_id = clear_invalid_fk(
            connection,
            RECORD_TABLE,
            &ID,
            "barcode_id",
            barcode_id,
            |c, id| BarcodeRowRepository::new(c).check_exists_by_id(id),
            true,
        )?;
        let location_id = clear_invalid_fk(
            connection,
            RECORD_TABLE,
            &ID,
            "location_id",
            location_ID,
            |c, id| LocationRowRepository::new(c).check_exists_by_id(id),
            true,
        )?;
        let item_variant_id = clear_invalid_fk(
            connection,
            RECORD_TABLE,
            &ID,
            "item_variant_id",
            item_variant_id,
            |c, id| ItemVariantRowRepository::new(c).check_exists_by_id(id),
            true,
        )?;
        let vvm_status_id = clear_invalid_fk(
            connection,
            RECORD_TABLE,
            &ID,
            "vvm_status_id",
            vvm_status_id,
            |c, id| VVMStatusRowRepository::new(c).check_exists_by_id(id),
            true,
        )?;

        let StockLineRowOmsFields {
            campaign_id,
            program_id,
            manufacture_date,
        } = oms_fields.unwrap_or_default();

        let campaign_id = clear_invalid_fk(
            connection,
            RECORD_TABLE,
            &ID,
            "campaign_id",
            campaign_id,
            |c, id| CampaignRowRepository::new(c).check_exists_by_id(id),
            true,
        )?;
        let program_id = clear_invalid_fk(
            connection,
            RECORD_TABLE,
            &ID,
            "program_id",
            program_id,
            |c, id| ProgramRowRepository::new(c).check_exists_by_id(id),
            true,
        )?;

        let result = StockLineRow {
            id: ID,
            store_id: store_ID,
            item_link_id: item_ID,
            location_id,
            batch,
            pack_size,
            cost_price_per_pack: cost_price,
            sell_price_per_pack: sell_price,
            available_number_of_packs: available,
            total_number_of_packs: quantity,
            expiry_date,
            on_hold: hold,
            note,
            supplier_id,
            barcode_id,
            item_variant_id,
            donor_id,
            manufacturer_id,
            vvm_status_id,
            campaign_id,
            program_id,
            manufacture_date,
            total_volume,
            volume_per_pack,
        };

        Ok(PullTranslateResult::upsert(result))
    }

    fn try_translate_to_upsert_sync_record(
        &self,
        connection: &StorageConnection,
        changelog: &ChangelogRow,
    ) -> Result<PushTranslateResult, anyhow::Error> {
        let Some(stock_line) = StockLineRepository::new(connection)
            .query_by_filter(
                StockLineFilter::new().id(EqualFilter::equal_to(changelog.record_id.to_string())),
                None,
            )?
            .pop()
        else {
            return Err(anyhow::anyhow!("Stock_line row not found"));
        };

        let StockLine {
            stock_line_row:
                StockLineRow {
                    id,
                    item_link_id: _,
                    store_id,
                    location_id,
                    batch,
                    pack_size,
                    cost_price_per_pack,
                    sell_price_per_pack,
                    available_number_of_packs,
                    total_number_of_packs,
                    expiry_date,
                    on_hold,
                    note,
                    supplier_id: _,
                    barcode_id,
                    item_variant_id,
                    donor_id: donor_link_id,
                    manufacturer_id,
                    vvm_status_id,
                    campaign_id,
                    program_id,
                    manufacture_date,
                    total_volume,
                    volume_per_pack,
                },
            item_row,
            supplier_name_row,
            ..
        } = stock_line;

        let oms_fields = Some(StockLineRowOmsFields {
            campaign_id,
            program_id,
            manufacture_date,
        });

        let legacy_row = LegacyStockLineRow {
            ID: id,
            store_ID: store_id,
            item_ID: item_row.id,
            batch,
            expiry_date,
            hold: on_hold,
            location_ID: location_id,
            pack_size,
            available: available_number_of_packs,
            quantity: total_number_of_packs,
            cost_price: cost_price_per_pack,
            sell_price: sell_price_per_pack,
            note,
            supplier_id: supplier_name_row.map(|supplier| supplier.id),
            barcode_id,
            item_variant_id,
            donor_id: donor_link_id,
            vvm_status_id,
            manufacturer_id,
            oms_fields,
            total_volume,
            volume_per_pack,
        };

        Ok(PushTranslateResult::upsert(
            changelog,
            self.table_name(),
            serde_json::to_value(legacy_row)?,
        ))
    }

    fn try_translate_to_delete_sync_record(
        &self,
        _: &StorageConnection,
        changelog: &ChangelogRow,
    ) -> Result<PushTranslateResult, anyhow::Error> {
        Ok(PushTranslateResult::delete(changelog, self.table_name()))
    }
}

#[cfg(test)]
mod tests {
    use crate::sync::{
        test::merge_helpers::{merge_all_item_links, merge_all_name_links},
        translations::ToSyncRecordTranslationType,
    };

    use super::*;
    use repository::{
        campaign_row::CampaignRow,
        mock::{MockData, MockDataInserts},
        system_log_row::{SystemLogRowRepository, SystemLogType},
        test_db::{setup_all, setup_all_with_data},
        ChangelogFilter, ChangelogRepository, ContextRow, ProgramRow, SyncAction,
    };
    use serde_json::json;

    #[actix_rt::test]
    async fn test_stock_line_translation() {
        use crate::sync::test::test_data::stock_line as test_data;
        let translator = StockLineTranslation {};

        // Pre-populate program_a and campaign_a (referenced by ITEM_LINE_1 in test data)
        // so that FK validation in the translator doesn't null them out.
        let (_, connection, _, _) = setup_all_with_data(
            "test_stock_line_translation",
            MockDataInserts::none(),
            MockData {
                contexts: vec![ContextRow {
                    id: "test_ctx".to_string(),
                    name: "test ctx".to_string(),
                }],
                programs: vec![ProgramRow {
                    id: "program_a".to_string(),
                    master_list_id: None,
                    name: "program_a".to_string(),
                    context_id: "test_ctx".to_string(),
                    is_immunisation: false,
                    elmis_code: None,
                    deleted_datetime: None,
                }],
                campaigns: vec![CampaignRow {
                    id: "campaign_a".to_string(),
                    name: "Campaign A".to_string(),
                    ..Default::default()
                }],
                ..Default::default()
            },
        )
        .await;

        for record in test_data::test_pull_upsert_records() {
            assert!(translator.should_translate_from_sync_record(&record.sync_buffer_row));
            let translation_result = translator
                .try_translate_from_upsert_sync_record(&connection, &record.sync_buffer_row)
                .unwrap();

            assert_eq!(translation_result, record.translated_record);
        }
    }

    /// When optional FKs reference records that don't exist, the translator should:
    ///  - null out each invalid FK on the translated row
    ///  - write a `system_log` row of type `SyncTranslationFkError` for each invalid FK
    /// so the integration upstream of this translator doesn't fail on FK constraint violations.
    #[actix_rt::test]
    async fn test_stock_line_clears_invalid_optional_fks_and_writes_system_log() {
        let translator = StockLineTranslation {};
        let (_, connection, _, _) = setup_all(
            "test_stock_line_clears_invalid_optional_fks_and_writes_system_log",
            MockDataInserts::none(),
        )
        .await;

        let sync_record = SyncBufferRow {
            table_name: "item_line".to_string(),
            record_id: "ITEM_LINE_FK_INVALID".to_string(),
            data: r#"{
                "ID": "ITEM_LINE_FK_INVALID",
                "store_ID": "store_a",
                "item_ID": "item_a",
                "available": 1.0,
                "barcodeID": "missing_barcode",
                "batch": "",
                "cost_price": 0,
                "expiry_date": "0000-00-00",
                "hold": false,
                "location_ID": "missing_location",
                "name_ID": "",
                "note": "",
                "pack_size": 1,
                "quantity": 1,
                "sell_price": 0,
                "total_volume": 0,
                "volume_per_pack": 0,
                "vvm_status_id": "missing_vvm",
                "om_item_variant_id": "missing_item_variant",
                "oms_fields": {
                    "campaign_id": "missing_campaign",
                    "program_id": "does_not_exist_program"
                }
            }"#
            .to_string(),
            action: SyncAction::Upsert,
            ..Default::default()
        };

        let result = translator
            .try_translate_from_upsert_sync_record(&connection, &sync_record)
            .unwrap();

        let expected = PullTranslateResult::upsert(StockLineRow {
            id: "ITEM_LINE_FK_INVALID".to_string(),
            store_id: "store_a".to_string(),
            item_link_id: "item_a".to_string(),
            location_id: None,
            batch: None,
            pack_size: 1.0,
            cost_price_per_pack: 0.0,
            sell_price_per_pack: 0.0,
            available_number_of_packs: 1.0,
            total_number_of_packs: 1.0,
            expiry_date: None,
            on_hold: false,
            note: None,
            supplier_id: None,
            barcode_id: None,
            item_variant_id: None,
            donor_id: None,
            manufacturer_id: None,
            manufacture_date: None,
            vvm_status_id: None,
            campaign_id: None,
            program_id: None,
            volume_per_pack: 0.0,
            total_volume: 0.0,
        });
        assert_eq!(result, expected);

        // One system_log entry per invalid FK
        let logs = SystemLogRowRepository::new(&connection)
            .find_all()
            .unwrap();
        let fk_errors: Vec<_> = logs
            .iter()
            .filter(|l| l.r#type == SystemLogType::SyncTranslationFkError && l.is_error)
            .collect();
        assert_eq!(fk_errors.len(), 6, "got {fk_errors:?}");
        let messages: String = fk_errors
            .iter()
            .filter_map(|l| l.message.as_deref())
            .collect::<Vec<_>>()
            .join("\n");
        for fk_field in [
            "barcode_id",
            "location_id",
            "item_variant_id",
            "vvm_status_id",
            "campaign_id",
            "program_id",
        ] {
            assert!(
                messages.contains(fk_field),
                "{}",
                format!("expected message to mention {fk_field}; got:\n{messages}")
            );
            assert!(
                messages.contains("ITEM_LINE_FK_INVALID"),
                "expected message to include the record id"
            );
        }
    }

    #[actix_rt::test]
    async fn test_stock_line_push_merged() {
        // The item_links_merged function will merge ALL items into item_a, so all stock_lines should have an item_id of "item_a" regardless of their original item_id.
        let (mock_data, connection, _, _) =
            setup_all("test_stock_line_push_link_merged", MockDataInserts::all()).await;

        merge_all_item_links(&connection, &mock_data).unwrap();
        merge_all_name_links(&connection, &mock_data).unwrap();

        let repo = ChangelogRepository::new(&connection);
        let changelogs = repo
            .changelogs(
                0,
                1_000_000,
                Some(ChangelogFilter::new().table_name(ChangelogTableName::StockLine.equal_to())),
            )
            .unwrap();

        let translator = StockLineTranslation {};
        for changelog in changelogs {
            // Translate and sort
            assert!(translator.should_translate_to_sync_record(
                &changelog,
                &ToSyncRecordTranslationType::PushToLegacyCentral
            ));
            let translated = translator
                .try_translate_to_upsert_sync_record(&connection, &changelog)
                .unwrap();

            assert!(matches!(translated, PushTranslateResult::PushRecord(_)));

            let PushTranslateResult::PushRecord(translated) = translated else {
                panic!("Test fail, should translate")
            };

            assert_eq!(translated[0].record.record_data["item_ID"], json!("item_a"));

            // Supplier ID can be null. We want to check if the non-null supplier_ids is "name_a".
            if translated[0].record.record_data["name_ID"] != json!(null) {
                assert_eq!(translated[0].record.record_data["name_ID"], json!("name_a"));
            }
        }
    }
}

File: server-go/internal/sync/translations/stockline.go
package translations

import (
	"database/sql"
	"encoding/json"
	"fmt"

	sq "github.com/Masterminds/squirrel"

	"github.com/msupply-foundation/open-msupply/server-go/internal/db"
	"github.com/msupply-foundation/open-msupply/server-go/internal/repository"
	"github.com/msupply-foundation/open-msupply/server-go/internal/sync/synctypes"
)

// legacyStockLineOmsFields mirrors translations::stock_line::StockLineRowOmsFields (the OMS
// extension fields carried inside the legacy record's "oms_fields" object).
type legacyStockLineOmsFields struct {
	CampaignID *string `json:"campaign_id"`
	ProgramID  *string `json:"program_id"`
}

// legacyStockLineRow mirrors translations::stock_line::LegacyStockLineRow (legacy table
// "item_line"). Optional strings are pointers so "" / absent become NULL.
type legacyStockLineRow struct {
	ID            string                    `json:"ID"`
	StoreID       string                    `json:"store_ID"`
	ItemID        string                    `json:"item_ID"`
	Batch         *string                   `json:"batch"`
	ExpiryDate    *string                   `json:"expiry_date"`
	Hold          bool                      `json:"hold"`
	LocationID    *string                   `json:"location_ID"`
	PackSize      float64                   `json:"pack_size"`
	Available     float64                   `json:"available"`
	Quantity      float64                   `json:"quantity"`
	CostPrice     float64                   `json:"cost_price"`
	SellPrice     float64                   `json:"sell_price"`
	Note          *string                   `json:"note"`
	SupplierID    *string                   `json:"name_ID"`
	BarcodeID     *string                   `json:"barcodeID"`
	ItemVariantID *string                   `json:"om_item_variant_id"`
	DonorID       *string                   `json:"donor_id"`
	VVMStatusID   *string                   `json:"vvm_status_id"`
	OmsFields     *legacyStockLineOmsFields `json:"oms_fields"`
	TotalVolume   float64                   `json:"total_volume"`
	VolumePerPack float64                   `json:"volume_per_pack"`
}

// StockLine is a tracked (changelog) remote-data entity. Legacy table "item_line".
type StockLine struct{ synctypes.BaseTranslation }

func (StockLine) TableNames() []string { return []string{"item_line"} }

func (StockLine) PullDependencies() []string {
	return []string{"item", "item_variant", "name", "store", "location", "barcode", "vvm_status", "campaign"}
}

func (StockLine) ChangelogType() (synctypes.ChangelogTableName, bool) {
	return synctypes.ChangelogStockLine, true
}

// ptrToNS maps a JSON optional string to a NullString, treating nil OR "" as NULL — mirroring
// util::sync_serde::empty_str_as_option_string used on every optional field of the legacy row.
func ptrToNS(p *string) sql.NullString {
	if p == nil || *p == "" {
		return sql.NullString{}
	}
	return sql.NullString{String: *p, Valid: true}
}

func (StockLine) TryTranslateFromUpsert(exec synctypes.Exec, d db.Dialect, row *synctypes.SyncBufferRow) (synctypes.PullTranslateResult, error) {
	var data legacyStockLineRow
	if err := json.Unmarshal([]byte(row.Data), &data); err != nil {
		return synctypes.NotMatched(), err
	}

	// Clear optional FKs whose referenced rows are missing (mirrors clear_invalid_fk).
	barcodeID, err := clearFK(exec, d, "barcode", ptrToNS(data.BarcodeID))
	if err != nil {
		return synctypes.NotMatched(), err
	}
	locationID, err := clearFK(exec, d, "location", ptrToNS(data.LocationID))
	if err != nil {
		return synctypes.NotMatched(), err
	}
	itemVariantID, err := clearFK(exec, d, "item_variant", ptrToNS(data.ItemVariantID))
	if err != nil {
		return synctypes.NotMatched(), err
	}
	vvmStatusID, err := clearFK(exec, d, "vvm_status", ptrToNS(data.VVMStatusID))
	if err != nil {
		return synctypes.NotMatched(), err
	}

	var campaign, program *string
	if data.OmsFields != nil {
		campaign, program = data.OmsFields.CampaignID, data.OmsFields.ProgramID
	}
	campaignID, err := clearFK(exec, d, "campaign", ptrToNS(campaign))
	if err != nil {
		return synctypes.NotMatched(), err
	}
	programID, err := clearFK(exec, d, "program", ptrToNS(program))
	if err != nil {
		return synctypes.NotMatched(), err
	}

	return synctypes.PullUpsert(repository.StockLineRow{
		ID:                     data.ID,
		StoreID:                data.StoreID,
		ItemLinkID:             data.ItemID,
		LocationID:             locationID,
		Batch:                  ptrToNS(data.Batch),
		PackSize:               data.PackSize,
		CostPricePerPack:       data.CostPrice,
		SellPricePerPack:       data.SellPrice,
		AvailableNumberOfPacks: data.Available,
		TotalNumberOfPacks:     data.Quantity,
		ExpiryDate:             zeroDateAsOption(deref(data.ExpiryDate)),
		OnHold:                 data.Hold,
		Note:                   ptrToNS(data.Note),
		BarcodeID:              barcodeID,
		ItemVariantID:          itemVariantID,
		VVMStatusID:            vvmStatusID,
		CampaignID:             campaignID,
		ProgramID:              programID,
		TotalVolume:            data.TotalVolume,
		VolumePerPack:          data.VolumePerPack,
		SupplierLinkID:         ptrToNS(data.SupplierID),
		DonorLinkID:            ptrToNS(data.DonorID),
	}), nil
}

func deref(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}

// TryTranslateToUpsert reads the stored stock line back (resolving item_link_id -> item_id and
// supplier_link_id -> name_id) and serializes it to the legacy "item_line" record.
func (s StockLine) TryTranslateToUpsert(exec synctypes.Exec, d db.Dialect, cl *synctypes.ChangelogRow) (synctypes.PushTranslateResult, error) {
	q := sq.Select(
		"sl.id", "sl.store_id", "il.item_id", "sl.batch", "sl.expiry_date", "sl.on_hold",
		"sl.location_id", "sl.pack_size", "sl.available_number_of_packs", "sl.total_number_of_packs",
		"sl.cost_price_per_pack", "sl.sell_price_per_pack", "sl.note", "snl.name_id",
		"sl.barcode_id", "sl.item_variant_id", "sl.donor_link_id", "sl.vvm_status_id",
		"sl.campaign_id", "sl.program_id", "sl.total_volume", "sl.volume_per_pack",
	).
		From("stock_line sl").
		Join("item_link il ON sl.item_link_id = il.id").
		LeftJoin("name_link snl ON sl.supplier_link_id = snl.id").
		Where(sq.Eq{"sl.id": cl.RecordID}).
		PlaceholderFormat(placeholderFor(d))

	var r legacyStockLineRow
	var batch, expiry, location, note, supplier, barcode, itemVariant, donor, vvm, campaign, program sql.NullString
	err := q.RunWith(exec).QueryRow().Scan(
		&r.ID, &r.StoreID, &r.ItemID, &batch, &expiry, &r.Hold, &location, &r.PackSize,
		&r.Available, &r.Quantity, &r.CostPrice, &r.SellPrice, &note, &supplier, &barcode,
		&itemVariant, &donor, &vvm, &campaign, &program, &r.TotalVolume, &r.VolumePerPack)
	if err == sql.ErrNoRows {
		return synctypes.PushNotMatchedResult(), fmt.Errorf("stock_line %s not found for push", cl.RecordID)
	}
	if err != nil {
		return synctypes.PushNotMatchedResult(), err
	}

	r.Batch, r.ExpiryDate, r.LocationID, r.Note = nsToPtr(batch), nsToPtr(expiry), nsToPtr(location), nsToPtr(note)
	r.SupplierID, r.BarcodeID, r.ItemVariantID, r.DonorID, r.VVMStatusID =
		nsToPtr(supplier), nsToPtr(barcode), nsToPtr(itemVariant), nsToPtr(donor), nsToPtr(vvm)
	r.OmsFields = &legacyStockLineOmsFields{CampaignID: nsToPtr(campaign), ProgramID: nsToPtr(program)}

	data, err := json.Marshal(r)
	if err != nil {
		return synctypes.PushNotMatchedResult(), err
	}
	return synctypes.PushUpsert(cl, s.TableNames()[0], data), nil
}

func (s StockLine) TryTranslateToDelete(_ synctypes.Exec, _ db.Dialect, cl *synctypes.ChangelogRow) (synctypes.PushTranslateResult, error) {
	return synctypes.PushDelete(cl, s.TableNames()[0]), nil
}
